// microsoft-cognitiveservices-speech-sdk/distrib/lib/src/sdk/AvatarSceneConfig.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AvatarSceneConfig = void 0;
/**
 * Defines the avatar scene configuration for controlling avatar positioning and orientation.
 * @class AvatarSceneConfig
 * Added in version 1.44.0
 *
 * @experimental This feature is experimental and might change or have limited support.
 */
class AvatarSceneConfig {
    /**
     * Creates and initializes an instance of this class with default values.
     * @constructor
     * @param {number} zoom - The zoom level (0-1, default 1.0).
     * @param {number} positionX - The horizontal position offset (-1 to 1, default 0.0).
     * @param {number} positionY - The vertical position offset (-1 to 1, default 0.0).
     * @param {number} rotationX - The rotation around the X axis in radians (default 0.0).
     * @param {number} rotationY - The rotation around the Y axis in radians (default 0.0).
     * @param {number} rotationZ - The rotation around the Z axis in radians (default 0.0).
     * @param {number} amplitude - The amplitude of the avatar's movements (0 to 1, default 1.0).
     */
    constructor(zoom = 1.0, positionX = 0.0, positionY = 0.0, rotationX = 0.0, rotationY = 0.0, rotationZ = 0.0, amplitude = 1.0) {
        this.zoom = zoom;
        this.positionX = positionX;
        this.positionY = positionY;
        this.rotationX = rotationX;
        this.rotationY = rotationY;
        this.rotationZ = rotationZ;
        this.amplitude = amplitude;
    }
}
exports.AvatarSceneConfig = AvatarSceneConfig;
